kamailio/users/config/kamailio.cfg

```
@@ -2314,6 +2297,47 @@ route[NATMANAGE] {
     }
 }
 
+route[SAVE_EXTPARTY] {
+    if ($var(is_from_inside) && $du != $null) {
+        # Calling to behind NAT UAC, save real address
+        $var(extparty) = ";extparty=sip:" + $(du{uri.host}) + ':' + $(du{uri.port});
+        xinfo("[$dlg_var(cidhash)] SAVE-EXTPARTY: add $var(extparty) to Record-Route\n");
+        add_rr_param("$var(extparty)");
+    }
+
+    if (!$var(is_from_inside) && !is_first_hop()) {
+        # Call from behind NAT UAC and not first hop (middle proxy), save real address
+        $var(extparty) = ";extparty=sip:" + $si + ':' + $sp;
+        xinfo("[$dlg_var(cidhash)] SAVE-EXTPARTY: add $var(extparty) to Record-Route\n");
+        add_rr_param("$var(extparty)");
+    }
```

Alternative implementation, although current one seems clear to me.

```suggestion
    $var(extparty) = "";
    
    if ($var(is_from_inside)) {
        if ($du != $null) {
            # Calling to behind NAT UAC, save real address
            $var(extparty) = ";extparty=sip:" + $(du{uri.host}) + ':' + $(du{uri.port});
        }
    } else {
        if (!is_first_hop()) {
            # Call from behind NAT UAC and not first hop (middle proxy), save real address
            $var(extparty) = ";extparty=sip:" + $si + ':' + $sp;
        }
    
    }

    if ($var(extparty)) {
        xinfo("[$dlg_var(cidhash)] SAVE-EXTPARTY: add $var(extparty) to Record-Route\n");
        add_rr_param("$var(extparty)");
    }
```